File: mm-swift/Sources/MetaMessage/Ir/ValueType.swift
import Foundation

public enum ValueType: UInt8, Codable {
    case unknown = 0

    case doc = 1
    case vec = 2
    case arr = 3
    case obj = 4
    case map = 5

    case str = 6
    case bytes = 7
    case bool = 8

    case i = 9
    case i8 = 10
    case i16 = 11
    case i32 = 12
    case i64 = 13
    case u = 14
    case u8 = 15
    case u16 = 16
    case u32 = 17
    case u64 = 18

    case f32 = 19
    case f64 = 20

    case bigint = 21
    case datetime = 22
    case date = 23
    case time = 24

    case uuid = 25
    case decimal = 26
    case ip = 27
    case url = 28
    case email = 29

    case enums = 30
    case media = 31

    public var stringValue: String {
        switch self {
        case .unknown: return "unknown"
        case .doc: return "doc"
        case .vec: return "vec"
        case .arr: return "arr"
        case .obj: return "obj"
        case .map: return "map"
        case .str: return "str"
        case .bytes: return "bytes"
        case .bool: return "bool"
        case .i: return "i"
        case .i8: return "i8"
        case .i16: return "i16"
        case .i32: return "i32"
        case .i64: return "i64"
        case .u: return "u"
        case .u8: return "u8"
        case .u16: return "u16"
        case .u32: return "u32"
        case .u64: return "u64"
        case .f32: return "f32"
        case .f64: return "f64"
        case .bigint: return "bigint"
        case .datetime: return "datetime"
        case .date: return "date"
        case .time: return "time"
        case .uuid: return "uuid"
        case .decimal: return "decimal"
        case .ip: return "ip"
        case .url: return "url"
        case .email: return "email"
        case .enums: return "enums"
        case .media: return "media"
        }
    }

    public static func parse(_ s: String) -> ValueType? {
        switch s.lowercased() {
        case "unknown": return .unknown
        case "doc": return .doc
        case "vec": return .vec
        case "arr", "array": return .arr
        case "obj": return .obj
        case "map": return .map
        case "str": return .str
        case "bytes": return .bytes
        case "bool": return .bool
        case "i": return .i
        case "i8": return .i8
        case "i16": return .i16
        case "i32": return .i32
        case "i64": return .i64
        case "u": return .u
        case "u8": return .u8
        case "u16": return .u16
        case "u32": return .u32
        case "u64": return .u64
        case "f32": return .f32
        case "f64": return .f64
        case "bigint": return .bigint
        case "datetime": return .datetime
        case "date": return .date
        case "time": return .time
        case "uuid": return .uuid
        case "decimal": return .decimal
        case "ip": return .ip
        case "url": return .url
        case "email": return .email
        case "enums": return .enums
        case "media": return .media
        default: return nil
        }
    }

    public var needsQuotes: Bool {
        switch self {
        case .str, .bytes, .datetime, .date, .time, .uuid, .ip, .url, .email, .enums, .unknown:
            return true
        default:
            return false
        }
    }
}

File: mm-cs/src/MetaMessage/Ir/ValueType.cs
namespace MetaMessage.Ir;

public enum ValueType
{
    Unknown,

    Doc,
    Vec,
    Arr,
    Obj,
    Map,

    Str,
    Bytes,
    Bool,

    I,
    I8,
    I16,
    I32,
    I64,
    U,
    U8,
    U16,
    U32,
    U64,

    F32,
    F64,

    Bigint,
    Datetime,
    Date,
    Time,

    Uuid,
    Decimal,
    Ip,
    Url,
    Email,

    Enums,

    Media
}

public static class ValueTypeConstants
{
    public const string UnknownStr = "unknown";
    public const string DocStr = "doc";
    public const string ArrStr = "arr";
    public const string VecStr = "vec";
    public const string ObjStr = "obj";
    public const string MapStr = "map";
    public const string StrStr = "str";
    public const string BytesStr = "bytes";
    public const string BoolStr = "bool";
    public const string IStr = "i";
    public const string I8Str = "i8";
    public const string I16Str = "i16";
    public const string I32Str = "i32";
    public const string I64Str = "i64";
    public const string UStr = "u";
    public const string U8Str = "u8";
    public const string U16Str = "u16";
    public const string U32Str = "u32";
    public const string U64Str = "u64";
    public const string F32Str = "f32";
    public const string F64Str = "f64";
    public const string BigintStr = "bigint";
    public const string DatetimeStr = "datetime";
    public const string DateStr = "date";
    public const string TimeStr = "time";
    public const string UuidStr = "uuid";
    public const string DecimalStr = "decimal";
    public const string IpStr = "ip";
    public const string UrlStr = "url";
    public const string EmailStr = "email";
    public const string EnumStr = "enums";
    public const string MediaStr = "media";
}

public static class ValueTypeExtensions
{
    private static readonly Dictionary<string, ValueType> StringToValueType = new()
    {
        { ValueTypeConstants.UnknownStr, ValueType.Unknown },
        { ValueTypeConstants.DocStr, ValueType.Doc },
        { ValueTypeConstants.ArrStr, ValueType.Arr },
        { ValueTypeConstants.VecStr, ValueType.Vec },
        { ValueTypeConstants.ObjStr, ValueType.Obj },
        { ValueTypeConstants.MapStr, ValueType.Map },
        { ValueTypeConstants.StrStr, ValueType.Str },
        { ValueTypeConstants.BytesStr, ValueType.Bytes },
        { ValueTypeConstants.BoolStr, ValueType.Bool },
        { ValueTypeConstants.IStr, ValueType.I },
        { ValueTypeConstants.I8Str, ValueType.I8 },
        { ValueTypeConstants.I16Str, ValueType.I16 },
        { ValueTypeConstants.I32Str, ValueType.I32 },
        { ValueTypeConstants.I64Str, ValueType.I64 },
        { ValueTypeConstants.UStr, ValueType.U },
        { ValueTypeConstants.U8Str, ValueType.U8 },
        { ValueTypeConstants.U16Str, ValueType.U16 },
        { ValueTypeConstants.U32Str, ValueType.U32 },
        { ValueTypeConstants.U64Str, ValueType.U64 },
        { ValueTypeConstants.F32Str, ValueType.F32 },
        { ValueTypeConstants.F64Str, ValueType.F64 },
        { ValueTypeConstants.BigintStr, ValueType.Bigint },
        { ValueTypeConstants.DatetimeStr, ValueType.Datetime },
        { ValueTypeConstants.DateStr, ValueType.Date },
        { ValueTypeConstants.TimeStr, ValueType.Time },
        { ValueTypeConstants.UuidStr, ValueType.Uuid },
        { ValueTypeConstants.DecimalStr, ValueType.Decimal },
        { ValueTypeConstants.IpStr, ValueType.Ip },
        { ValueTypeConstants.UrlStr, ValueType.Url },
        { ValueTypeConstants.EmailStr, ValueType.Email },
        { ValueTypeConstants.EnumStr, ValueType.Enums },
        { ValueTypeConstants.MediaStr, ValueType.Media }
    };

    public static string ToTypeString(this ValueType vt)
    {
        return vt switch
        {
            ValueType.Unknown => ValueTypeConstants.UnknownStr,
            ValueType.Doc => ValueTypeConstants.DocStr,
            ValueType.Arr => ValueTypeConstants.ArrStr,
            ValueType.Vec => ValueTypeConstants.VecStr,
            ValueType.Obj => ValueTypeConstants.ObjStr,
            ValueType.Map => ValueTypeConstants.MapStr,
            ValueType.Str => ValueTypeConstants.StrStr,
            ValueType.Bytes => ValueTypeConstants.BytesStr,
            ValueType.Bool => ValueTypeConstants.BoolStr,
            ValueType.I => ValueTypeConstants.IStr,
            ValueType.I8 => ValueTypeConstants.I8Str,
            ValueType.I16 => ValueTypeConstants.I16Str,
            ValueType.I32 => ValueTypeConstants.I32Str,
            ValueType.I64 => ValueTypeConstants.I64Str,
            ValueType.U => ValueTypeConstants.UStr,
            ValueType.U8 => ValueTypeConstants.U8Str,
            ValueType.U16 => ValueTypeConstants.U16Str,
            ValueType.U32 => ValueTypeConstants.U32Str,
            ValueType.U64 => ValueTypeConstants.U64Str,
            ValueType.F32 => ValueTypeConstants.F32Str,
            ValueType.F64 => ValueTypeConstants.F64Str,
            ValueType.Bigint => ValueTypeConstants.BigintStr,
            ValueType.Datetime => ValueTypeConstants.DatetimeStr,
            ValueType.Date => ValueTypeConstants.DateStr,
            ValueType.Time => ValueTypeConstants.TimeStr,
            ValueType.Uuid => ValueTypeConstants.UuidStr,
            ValueType.Decimal => ValueTypeConstants.DecimalStr,
            ValueType.Ip => ValueTypeConstants.IpStr,
            ValueType.Url => ValueTypeConstants.UrlStr,
            ValueType.Email => ValueTypeConstants.EmailStr,
            ValueType.Enums => ValueTypeConstants.EnumStr,
            ValueType.Media => ValueTypeConstants.MediaStr,
            _ => ValueTypeConstants.UnknownStr
        };
    }

    public static ValueType ParseValueType(string s)
    {
        s = s.ToLower();
        if (StringToValueType.TryGetValue(s, out var vt))
        {
            return vt;
        }
        return ValueType.Unknown;
    }

    public static bool NeedsQuotes(this ValueType vt)
    {
        return vt switch
        {
            ValueType.Str => true,
            ValueType.Bytes => true,
            ValueType.Datetime => true,
            ValueType.Date => true,
            ValueType.Time => true,
            ValueType.Uuid => true,
            ValueType.Ip => true,
            ValueType.Url => true,
            ValueType.Email => true,
            ValueType.Enums => true,
            _ => false
        };
    }
}